Program.cs

```diff
@@ -1,41 +1,53 @@
-
-    var treasureCounter = 0;
-    Console.Clear();
-    DisplayGrid(grid);
-
-    ConsoleKeyInfo keyInfo;
-    while ((keyInfo = Console.ReadKey(true)).Key != ConsoleKey.Escape)
-    {
-        (playerRow, playerColumn, treasureCounter) =
-            MovePlayer(grid, playerRow, playerColumn, keyInfo, treasureCounter);
-        (monsterRow, monsterColumn) = MoveMonster(grid, playerRow, playerColumn, monsterRow, monsterColumn);
+    var gameRunning = true;
 
+    while (gameRunning)
+    {
+        var (gridRows, gridColumns) = GetGridSize();
+        var grid = InitializeGrid(gridRows, gridColumns);
+        var numberOfTreasures = (int)(gridRows * gridColumns * 0.1);
+        var numberOfObstacles = (int)(gridRows * gridColumns * 0.2);
+        var (playerRow, playerColumn) = PlacePlayer(grid);
+        var (monsterRow, monsterColumn) = PlaceMonster(grid);
+
+        var treasureCounter = 0;
         Console.Clear();
         DisplayGrid(grid);
-        Console.WriteLine("Number of Treasures: " + numberOfTreasures);
-        Console.WriteLine("Treasure Counter: " + treasureCounter);
+        PlaceItems(numberOfTreasures, '$', grid);
+        PlaceItems(numberOfObstacles, '|', grid);
 
-        if (treasureCounter == numberOfTreasures)
-        {
-            Console.Clear();
-            Console.WriteLine($"You found all {numberOfTreasures} treasures and win the game!");
-            break;
-        }
+        ConsoleKeyInfo keyInfo;
+        var gameActive = true;
 
-        if (monsterRow == playerRow && monsterColumn == playerColumn)
+        while (gameActive && (keyInfo = Console.ReadKey(true)).Key != ConsoleKey.Escape)
         {
+            (playerRow, playerColumn, treasureCounter) =
+                MovePlayer(grid, playerRow, playerColumn, keyInfo, treasureCounter);
+            (monsterRow, monsterColumn) = MoveMonster(grid, playerRow, playerColumn, monsterRow, monsterColumn);
+
             Console.Clear();
-            Console.WriteLine("The monster caught you! Game over.");
-            TreasureHunt();
+            DisplayGrid(grid);
+            Console.WriteLine("Number of Treasures: " + numberOfTreasures);
+            Console.WriteLine("Treasure Counter: " + treasureCounter);
+
+            if (treasureCounter == numberOfTreasures)
+            {
+                Console.Clear();
+                Console.WriteLine($"You found all {numberOfTreasures} treasures and win the game!");
+                gameActive = false;
+            }
+            else if (monsterRow == playerRow && monsterColumn == playerColumn)
+            {
+                Console.Clear();
+                Console.WriteLine("The monster caught you! Game over.");
+                gameActive = false;
+            }
         }
+
+        Console.WriteLine("Press 'R' to restart the game or any other key to exit.");
+        var restartKey = Console.ReadKey(true).Key;
+        if (restartKey != ConsoleKey.R) gameRunning = false;
     }
 }
 
```

Comment:
Ad. GetGridSize() - you should check the length of the array after using input.Split. User input can't be trusted.

Reply:
The method signature should also indicate what it returns - e.g. `static (int Height, int Width) GetGridSize()`